templates/pkg/resource/manager.go.tpl

```diff
@@ -156,6 +162,69 @@ func (rm *resourceManager) ARNFromName(name string) string {
 	)
 }
 
+// LateInitialize returns an acktypes.AWSResource after setting the late initialized
+// fields from the readOne call. This method will initialize the optional fields
+// which were not provided by the k8s user but were defaulted by the AWS service.
+// If there are no such fields to be initialized, the returned object is similar to
+// object passed in the parameter.
+func (rm *resourceManager) LateInitialize(
+	ctx context.Context,
+	res acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+	rlog := ackrtlog.FromContext(ctx)
+{{- if $hookCode := Hook .CRD "late_initialize_pre_read_one" }}
+{{ $hookCode }}
+{{- end }}
+	// If there are no fields to late initialize, do nothing
+	if len(lateInitializeFieldNames) == 0 {
+		rlog.Debug("no late initialization required.")
+		return res, nil
+	}
+	r := rm.concreteResource(res)
+	if r.ko == nil {
+		// Should never happen... if it does, it's buggy code.
+		panic("resource manager's LateInitialize() method received resource with nil CR object")
+	}
+	koWithDefaults := r.ko.DeepCopy()
+	latestWithDefaults := &resource{koWithDefaults}
+	lateInitConditionReason := ""
+	lateInitConditionMessage := ""
+
+	observed, err := rm.sdkFind(ctx, latestWithDefaults)
+	if err != nil {
+		lateInitConditionMessage = "Unable to complete Read operation required for late initialization"
+		lateInitConditionReason = "Late Initialization Failure"
+		ackcondition.SetLateInitialized(latestWithDefaults, corev1.ConditionFalse, &lateInitConditionMessage, &lateInitConditionReason)
+		return latestWithDefaults, ackrequeue.NeededAfter(err, time.Duration(0)*time.Second)
+	}
+	observedKo := observed.ko
+{{ GoCodeLateInitializeFromReadOne .CRD "observedKo" "koWithDefaults" 1 }}
+	incompleteInitialization := rm.incompleteLateInitialization(latestWithDefaults)
+	if incompleteInitialization {
+		// Add the condition with LateInitialized=False
+		lateInitConditionMessage = "Late initialization did not complete, requeuing with delay of 5 seconds"
+		lateInitConditionReason = "Delayed Late Initialization"
+		ackcondition.SetLateInitialized(latestWithDefaults, corev1.ConditionFalse, &lateInitConditionMessage, &lateInitConditionReason)
+		return latestWithDefaults, ackrequeue.NeededAfter(nil, time.Duration(5)*time.Second)
+	}
```

Comment:
I'm struggling to understand why we need to construct another resource in `GoCodeLateInitializeFromReadOne` at all, since the `latest` returned from `rm.ReadOne` would have all the Spec and Status fields defaulted properly already.

We could then simply compare the original (`res`) with the `latest` resource using `newDelta()` and that would give us the fields that had been server-side defaulted.

Something like this:

```go
{{- if $hookCode := Hook .CRD "late_initialize_post_read_one" }}
{{ $hookCode }}
{{- end }}
	delta := newResourceDelta(res, latest)
	if delta.DifferentAt("Spec") {
	    msg := fmt.Sprintf(
	        "Fields were late-initialized (server-side defaults): %s",
	        delta.Differences,
	    )
	    ackcondition.SetLateInitialized(latest, corev1.ConditionTrue, &msg, &lateInitConditionReason)
	}
	return latest, nil
}
```

Reply:
> I'm struggling to understand why we need to construct another resource in GoCodeLateInitializeFromReadOne at all, since the latest returned from rm.ReadOne would have all the Spec and Status fields defaulted properly already.

It is because the latest from ReadOne will not only have fields that are late initialized, It will have fields that were overriden by the server as well. Hence in the method `LateInitializeFromReadOne` , we only update the fields which were not specified by the user and added by the service.

> We could then simply compare the original (res) with the latest resource using newDelta() and that would give us the fields that had been server-side defaulted.

Excellent suggestion but I see some shortcomings in current delta code. It does not have complete support for nested Maps("..") i.e. when value of Map is struct. It may be a bigger effort to add support in delta but it is very easy to just do comparison check inside `rm.incompleteLateInitialization` even for nested fields.
I can add a TODO here to reuse Delta function when feasible in future.
Also, this simplified code does not handle case of partial late initialization where for example only one of two late initialized fields is initialized and other will be present in future readOne output.

Once we have agreement over this, I will make the code changes incorporation both of your comments.